### PR TITLE
Support #1619 - MTD vrs FTD builds, and Makefile.am If/Else/Endif issues

### PR DESCRIPTION
**Pull request by DuaneEllis-TI**
As discussed in #1467 (Changes to Support IDEs)

These problems are intertwined and share a number of common traits and are being addressed together in a single commit, sadly I don't see a way to break this up.

PROBLEM 1
In src/core/Makefile.am - lists numerous IF/ELSE/ENDIF which choose which files are added or not added to a specific library. While IDEs can provide some level of exclusion, they are not that flexible.

PROBLEM 2
Need both --enable-cli-ftd --enable-cli-mtd, and likewise --enable-mtd

Need granularity such that CLI can be built with MTD only, And NCP can be built using MTD or FTD
Current ./configure solution is: FTD for both, or FTD for Neither

(there are no details for this item below)

PROBLEM 3
Linking the CLI application with the CLI library, and NCP with the NCP library

Currently when the CLI (or NCP) library is built - it is built using only one of the two internal header configurations, ie: FTD - with all of the structure/class elements, or MTD with various structure/class elements removed.

This causes problems with link time optimization and static code analysis.

PROBLEM 4
Appvayor.

I do not have access to the version of Visual Studio used here, I have Visual Studio 2015, but the project appears to use an older version of Visual Studio, my experience has been: VS2015 will upgrade the projects... this is a one-way process

if someone can advise, it would be helpful.

DETAILS FOLLOW

================================================

PROBLEM 1 - details

a) Makefile.am files should not longer use If/Else/Endif type constructs to add/remove files from a build.
Instead, the makefile should generally compile everything.

For example some IDEs - such as Eclipse CDT based IDEs - compile all files ".c" and ".cpp" and are of the view - if the file does not belong, or should not be compiled then it should not be present in the directory. While other IDEs have other schemes, in IDE land - there does not exist a single common way to exclude/include files based upon the build configuration.

Solution: Where required, appropriate #if/#else/#endif type statements need to be added to various source files so that the entire quote "translation unit" - effectively becomes a pseudo-empty translation unit.

A number of files did not have, or do not #include "openthread-config.h" - which is required to support the #if/#else/#endif described above, this was added to a number of files.

Result: A number of places additional tests must be applied to the conditional.

For example in several places, the #if SOMETHING becomes: #if SOMETHING && OPENTHREAD_FTD

PROBLEM 2 Details

This is described above, not expanded here.

PROBLEM 3 Details

When the NCP or CLI library is built, it uses the internal thread headers, and thus structures and class members.

When the NCP or CLI application is linked, the openthread library effectively uses a different set of header files (either MTD or FTD configuration) - the net result is these internal structures are different.

What is needed is the CLI library must be built in FTD mode, and MTD mode so that it matches the openthread library, these differences are flagged as errors by compilers and static code analysis tools that do "link-time-optimization"

Thus, in all cases we have added two command line defines:

When building anything-FTD, "-DOPENTHREAD_FTD=1" must be defined
When building anything-MTD, "-DOPENTHREAD_MTD=1" must be defined

PROBLEM 4 - No access to Visual Studio OLD?

I have Visual Studio 2015 - I do not have older versions, I cannot make any of the required changes to support the above for appvayor.

I believe the change is simple, because I believe the Windows solution only builds the FTD configuration.

In the library build projects add: -DOPENTHREAD_FTD=1

### CODE REVIEW

File: .travis/script.sh [hunk 1]
```diff
@@ -144,7 +144,7 @@ set -x
 
 [ $BUILD_TARGET != posix-ncp-spi ] || {
     ./bootstrap || die
-    make -f examples/Makefile-posix check configure_OPTIONS="--enable-ftd --enable-ncp=spi --with-examples=posix --with-platform-info=POSIX" || die
+    make -f examples/Makefile-posix check configure_OPTIONS="--enable-ftd --enable-ncp-app=all --enable-cli-app=all --with-ncp-bus=spi --with-examples=posix --with-platform-info=POSIX" || die
```
Comment (jwhui) [inline .travis/script.sh]: Remove the `--enable-cli-app=all` option.  This build check is not intended to run the python test scripts.

**Comment by jwhui** [comment]
FWIW, I've been editing the project files in [etc/visual-studio](https://github.com/openthread/openthread/tree/master/etc/visual-studio) directly using a text editor.  You'll probably want to add `OPENTHREAD_FTD` to the `PreprocessorDefinitions` blocks.

**Comment by nibanks** [comment]
Or you can add it to [openthread-windows-config.h](https://github.com/openthread/openthread/blob/master/include/openthread-windows-config.h).

File: .travis/script.sh (hunk 1 above)
Comment (DuaneEllis-TI) [inline .travis/script.sh]: Done, dealing with auto-merge issues now.

File: etc/visual-studio/libopenthread-cli-windows.vcxproj [hunk 2]
```diff
@@ -42,6 +42,7 @@
         OPENTHREAD_CONFIG_FILE="openthread-windows-config.h";
         OTDLL;
         OTBUILD;
+	OPENTHREAD_FTD=1;
```
Comment (nibanks) [inline etc/visual-studio/libopenthread-cli-windows.vcxproj]: Please make sure the spacing it correct.

Comment (DuaneEllis-TI) [inline etc/visual-studio/libopenthread-cli-windows.vcxproj]: Just pushed that (I had tabs, file is actually spaces)  Tedious.

**Comment by DuaneEllis-TI** [comment]
Nick Banks (@nibanks) 

Sadly - I had to edit the vcxproj files ... 

Another item is this which is different from openthread.

In openthreads case, we ./configure the NCP to be either SPI or UART - it does not build both versions, this is done in the 'openthread-config.h' file, today the choice is binary - either SPI or UART 

In contrast, the windows builds do both UART and SPI at the same time, I could not duplicate the config.h method because this would require two different openthread-config.h files for windows.
Thus, in those cases I had to add the #defines on the commandline for the NCP library build.

Please take a look.

Thanks for your help.

**Comment by darconeous** [comment]
Technically, the `--with-*`/`--without-*` options are supposed to be used for determining which external dependencies to use. It is the `--enable-*`/`--disable-*` that is supposed to be used for adjusting features. 

From [Autotools Mythbuster](https://autotools.io/autoconf/arguments.html):

> There are three types of options (or properly arguments ) that can be added to the configure script:
>
> * `--enable-*`/`--disable-*` arguments  
>     The arguments starting with `--enable-` prefix are usually used to enable features of the program. They usually add or remove dependencies only if they are needed for that particular feature being enabled or disabled.
>* `--with-*`/`--without-*` arguments  
>    The arguments starting with `--with-` prefix are usually used to add or remove dependencies on external projects. These might add or remove features from the project.
> * environment variables  
>    Environment variables that are used by the configure script should also be declared as arguments; their use will be explained below in detail.
> 
> The first two kinds of parameters differ just for the displayed name and from the macro used, but are in effect handled mostly in the same way. They both are actually used to pass variables, in the form of `--(enable|with)-foo=bar` and both provide defaults for when the variable is omitted (the value yes for `--enable` and `--with` and the value no for `--disable` and `--without`).
> 
> While there is no technical difference between the two, it's helpful for both users and distribution to follow the indications given above about the use of the two parameters' kind. This allows to identify exactly what the parameters are used for.

Using the command-line environment variable to specify the bus type seems like the most natural way to go about this, but I'm fairly flexible.

**Comment by DuaneEllis-TI** [comment]
My point was more something else.

There are 4 variants of the NCP application.

   ncp-{MTD|FTD}-{UART|SPI}

In ./configure land - the choice of UART/SPI - is at ./configure time, and 0, 1 or 2 variants (ie: None, MTD, FTD, or both/all)

In contrast, windows has taken the other approach,  Only build FTD, and build both  UART and SPI

The header files are not setup in the windows flavor. The question is, should we adjust? or not.

**Comment by nibanks** [comment]
For Windows, as far as building NCP, there isn't any product scenario that actually requires it. It is more just for ensuring build support/coverage and used for unit test library. For all other features, since Windows doesn't really have any processing or memory restrictions compared to a real device, so we try to enable every feature that makes sense. There is no need to make it configurable.

File: etc/visual-studio/UnitTests.vcxproj [hunk 3]
```diff
@@ -50,6 +50,7 @@
         MBEDTLS_CONFIG_FILE="mbedtls-config.h";
         OPENTHREAD_CONFIG_FILE="openthread-windows-config.h";
         OPENTHREAD_MULTIPLE_INSTANCE;
+        OPENTHREAD_FTD=1;
```
Comment (nibanks) [inline etc/visual-studio/UnitTests.vcxproj]: Looks like you could put the `OPENTHREAD_FTD=1` in the header file and just put the SPI and UART defines in the project files that need them. I don't care too much either way though.

Comment (DuaneEllis-TI) [inline etc/visual-studio/UnitTests.vcxproj]: I actually tried putting it in the header first, but it did not work.

There are a few combinations of include file order where 'openthread-config.h' not the first file being included, and ... because of the domino-chain of includes it did not work.

You can see what I mean if you move it back to the header.
That's a different problem that should be solved on a different PR/ticket.

**Comment by DuaneEllis-TI** [comment]
Q: What is blocking this from being merged?
Do I need to address anything?

**Comment by darconeous** [comment]
My earlier comment was addressing the use of `--with-ncp-bus=uart` and `--with-ncp-bus=spi`, which isn't technically what the `--with-*`/`--without-*` arguments are intended to be used for (see my previous comment for details). The current usage is inconsistent with how those types of arguments are intended to be used.

But I'm not going to recommend we hold back for that. We already aren't super consistent to begin with.

I'm going to give this another review pass right now, should take a few minutes.

**Comment by DuaneEllis-TI** [comment]
FYI - the "--with-ncp-bus" was per Jonhui's suggestion in a different discussion.

File: etc/visual-studio/libopenthread-windows.vcxproj [hunk 4]
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (darconeous) [inline etc/visual-studio/libopenthread-windows.vcxproj]: Any idea what changed on this line? Was it a UTF format marker prefix or something?

Comment (DuaneEllis-TI) [inline etc/visual-studio/libopenthread-windows.vcxproj]: yea, it seems there is/was a byte order mark <U+FEFF> that my editor must have munched/dropped....
And - the NEWLINE (or not) at the end of the file.

Ugh.

Comment (DuaneEllis-TI) [inline etc/visual-studio/libopenthread-windows.vcxproj]: So I fixed these, the diff now shows they are the same as before.

Interestingly, some of the files have a BOM - some do not (mostly they do)
Most do not have a final LF, some do (mostly they do not)

Not sure which editor did that to me... Ugh.

File: examples/apps/ncp/Makefile.am [hunk 5]
```diff
@@ -134,18 +137,19 @@ ot_ncp_ftd_LDFLAGS                                                     = \
 ot_ncp_ftd_SOURCES                                                     = \
     $(SOURCES_COMMON)                                                    \
     $(NULL)
-endif # OPENTHREAD_ENABLE_FTD
 
+if  OPENTHREAD_ENABLE_NCP_MTD
```
Comment (jwhui) [inline examples/apps/ncp/Makefile.am]: nit: extra space after `if`

File: src/cli/Makefile.am [hunk 6]
```diff
@@ -28,40 +28,136 @@
...
+    libopenthread-cli-ftd.a           \
+    $(NULL)
 
-libopenthread_cli_a_CPPFLAGS =        \
-    -I$(top_srcdir)/include           \
-    -I$(top_srcdir)/src               \
-    -I$(top_srcdir)/src/core          \
-    $(OPENTHREAD_TARGET_DEFINES)      \
-    $(MISSING_CPPFLAGS)               \
+CPPFLAGS_COMMON=                        \
```
Comment (jwhui) [inline src/cli/Makefile.am]: nit: missing space

File: src/cli/Makefile.am [hunk 7]
```diff
@@ -28,40 +28,136 @@
...
+    -I$(top_srcdir)/include             \
```
Comment (jwhui) [inline src/cli/Makefile.am]: nit: align `\` with rest of file

File: src/cli/cli.cpp [hunk 8]
```diff
@@ -2704,7 +2704,7 @@ void Interpreter::HandlePanIdConflict(uint16_t aPanId, uint32_t aChannelMask)
     sServer->OutputFormat("Conflict: %04x, %08x\r\n", aPanId, aChannelMask);
 }
 
-#endif  // OPENTHREAD_ENABLE_COMMISSIONER
+#endif  //  OPENTHREAD_ENABLE_COMMISSIONER && OPENTHREAD_FTD
```
Comment (jwhui) [inline src/cli/cli.cpp]: nit: extra space after `//`

File: src/cli/cli_coap.cpp [hunk 9]
```diff
@@ -31,11 +31,16 @@
  *   This file implements a simple CLI coap server and client.
  */
 
+
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: nit: extra newline

File: src/cli/cli_coap.cpp [hunk 10]
```diff
@@ -31,11 +31,16 @@
...
 #include <ctype.h>
 #include <cli/cli.hpp>
+
+#if OPENTHREAD_ENABLE_APPLICATION_COAP
+
 #include <cli/cli_coap.hpp>
 #include <coap/coap_header.hpp>
 
+
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: nit: extra newline

File: etc/visual-studio/libopenthread-ncp-uart.vcxproj [hunk 11]
```diff
@@ -73,4 +76,4 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets">
   </ImportGroup>
-</Project>
+</Project>
```
Comment (DuaneEllis-TI) [inline etc/visual-studio/libopenthread-ncp-uart.vcxproj]: You might want a newline, but this is not how the files are generated from Visual Studio.

Comment (jwhui) [inline etc/visual-studio/libopenthread-ncp-uart.vcxproj]: This PR actually removes the newline.  Are you using Visual Studio to edit these files?

File: examples/apps/ncp/Makefile.am (hunk 5 above)
Comment (DuaneEllis-TI) [inline examples/apps/ncp/Makefile.am]: done

File: src/cli/Makefile.am (hunk 6 above)
Comment (DuaneEllis-TI) [inline src/cli/Makefile.am]: done

File: src/cli/Makefile.am (hunk 7 above)
Comment (DuaneEllis-TI) [inline src/cli/Makefile.am]: done

File: src/cli/cli.cpp (hunk 8 above)
Comment (DuaneEllis-TI) [inline src/cli/cli.cpp]: done

File: src/cli/cli_coap.cpp (hunk 9 above)
Comment (DuaneEllis-TI) [inline src/cli/cli_coap.cpp]: done

File: src/cli/cli_coap.cpp (hunk 10 above)
Comment (DuaneEllis-TI) [inline src/cli/cli_coap.cpp]: done

File: etc/visual-studio/libopenthread-ncp-uart.vcxproj (hunk 11 above)
Comment (DuaneEllis-TI) [inline etc/visual-studio/libopenthread-ncp-uart.vcxproj]: yea, confusing:

I learned a new word here.   https://octicons.github.com/icon/no-newline/

Take a look at this earlier commit etc/visual-studio/libopenthread-ncp-uart.vcxproj

Question: What happened at line 1, and line 73/75? 

Answer: Line 1 - contained a byte order mark (U+FFEF), this is the commit that the byte order mark was removed.

Answer: Line 73/75 - *DID* not have newline, it was added during this commit.

Note: My commit does not *NOT* have the byte order mark, (this is something that @darconeous  noticed... in my commits, I fixed these.

Now fast forward to the current problem,

Compare my two GIT HUB commits, the only change is the octicon .. moved from one line to the other.
My change, adding/removing the terminal newline.

Surely I cannot delete what is on Github...  But Github says, in one case - the old file is missing a terminal newline, and in the other, commit the old file has a terminal newline.

I wonder if this is a github bug in the rendering/display code for bitbucket.

https://github.com/openthread/openthread/commit/9301cf12d7b41ec5970f25f7e2529f8a9f894949

https://github.com/openthread/openthread/commit/9301cf12d7b41ec5970f25f7e2529f8a9f894949

I can make this any way you want, - my experience is this:

a)  Visual studio would like - but does not require a U+FFEF byte order mark.

b)  Visual studio does not require, and seems to ignore a terminal newline.

Visual studio seems to follow Postal's Law.

Comment (jwhui) [inline etc/visual-studio/libopenthread-ncp-uart.vcxproj]: Thanks for looking into this.  I'm fine with accepting this PR as is.  Just wanted to make sure we weren't making unnecessary changes.

**Comment by jwhui** [comment]
@DuaneEllis-TI, thanks for these enhancements to bring boarder toolchain support!
